### DISCUSSION TITLE
XRPL Currency Codes: if we could design them today, from scratch, what would we specify?

### DISCUSSION THREAD

**Discussion by sappenin** [XRPLF/XRPL-Standards#128] [XLS Ideas (pre standard proposal)]
Does anyone have feedback around the question above? If we had to redo [Currency Codes](https://xrpl.org/currency-formats.html#currency-codes), would we remove, change or keep anything? If so, what? 

This discussion is mainly prompted by my work on [CFTs](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-CFTs), where I have received some very good feedback (which is prompting this broader discussion). In particular, I found this feedback to be insightful:

> I think the asset code description has too much "guidance" and not enough hard rules. If we really think this is useful guidance we should enforce it. If necessary, the rules can be relaxed later if there's a good reason. If the rules are very loose initially, then it's very difficult to tighten them later. Loose rules for the trust line asset code led to some awkward situations with abusive "issuers" early on.

Some axes to consider around Currency Codes:

1. Content restrictions? Should we allow any type of binary? Unicode? Or only ASCII? Some subset of ASCII?
2. Length restrictions? (e.g., see [here](https://github.com/MetaMask/metamask-extension/issues/9243) for why > 10 bytes has helped other communities).
3. How should these be rendered? (e.g., browser, app, CLI, etc).
4. Should we consider some sort of prefix scheme (like [this one](https://github.com/XRPLF/XRPL-Standards/discussions/53))?

**Comment by intelliot** [comment]
I would like to read a compare/contrast with the ERC-1155 Metadata URI JSON Schema. 

A currency code might not be necessary (and could be counterproductive) - consider NFTs, semi-fungible tokens, and other tokenizations. 

> The symbol function (found in the ERC-20 and ERC-721 standards) was not included as we do not believe this is a globally useful piece of data to identify a generic virtual item / asset and are also prone to collisions. Short-hand symbols are used in tickers and currency trading, but they aren’t as useful outside of that space.

https://eips.ethereum.org/EIPS/eip-1155

In practice, the ERC-1155 I have looked at still use name and symbol, but it's part of the more flexible and generalized JSON Metadata. 

That being said, I believe adoption to date of ERC-1155 has been relatively low. MetaMask does not yet support sending ERC-1155 tokens.

**Reply by sappenin** [reply]
Excellent points @intelliot, I hadn't considered that kind of design - looking at the JSON schema it's occurring to me that (1) there are other aspects of a "currency" or "asset" that especially impact off-chain things like UI/UX (e.g., [an image](https://eips.ethereum.org/EIPS/eip-1155#metadata)); (2) it seems like this meta-information doesn't necessarily need to live "on chain" since much of it is used off-chain.

[a comment by an automated account is not shown]
